Directly print PDFs in Chromium (+PDF editing functionality)

Rob--W opened the pull request:
This PR adds a minimalistic and low-level `PDFDataWriter` class to PDF.js, which allows us to (incrementally) update PDFs. With this new class, I add print instructions to the PDF, which significantly improves the printing performance in Chromium-based browsers with a built-in PDF previewer (so, basically every Chromium browser AFAIK).

I suggest to separately review each commit, instead of checking the combined diff of the PR, because each commit is focused on a specific feature and has a detailed commit message.

To verify that the whole works as expected:
1. `node make chromium`
2. Open Chrome, visit chrome://extensions and load the unpacked extension from build/chromium/.
3. Open any PDF (preferably one with 10+ pages) with Chrome.
4. Focus the page and press Ctrl + P.
5. A print preview should immediately be displayed, without rendering all pages upfront.
6. Close the preview and click on the print button (or Ctr+P).
7. Same as 5.

and
1. `node make generic`
2. `node make server` (to start the server).
3. Visit localhost:8888 and open a PDF
4. See step 4 - 7 of the previous list.

and
1. Visit `http://localhost:8888/test/unit/unit_test.html?spec=pdf_data_writer` and verify that all tests are green.

Rob--W committed:
Add PDFDataWriter class to generate PDF data

The PDFDataWriter class is a low-level class that supports the raw types
as defined by section 7.3 and 7.5 of the PDF 1.7 spec. It is designed to
support incremental updates in PDF.js.

The immediate use case for developing this class is to add a printing
instruction to the PDF for Chrome, but the functionality can also be
re-used to implement things like persistent annotations.

The code is not referenced in the worker loader yet, because at this
point there is no code that uses the writer. The class has fully been
tested, including a full example of using the class to create a PDF from
scratch (see 'should be able to generate a valid PDF file' in
test/unit/pdf_data_writer_spec.js and test/pdfs/written-with-pdfjs.pdf).

Rob--W committed:
Route all print requests through one function

This commit does not change existing behavior, the refactor enables the
implementation of a custom print handler in a later commit.

The viewer code should get the opportunity to handle the print request
before the mozPrintCallback polyfill intercepts Ctrl+P, so the
useCapture flag in mozPrintCallback has been set to false.

I verified that changing to useCapture=false does not change the
behavior of browsers with respect to the ability to prevent the print
dialog from being opened. In all cases, with and without the useCapture
flag, the following happens when the keydown event listener uses
event.preventDefault() upon Ctrl+P:

- Chrome 31, 43: The print dialog is suppressed.
- Firefox 3.6, 40: The print dialog is suppressed.
- Opera 31: The print dialog is suppressed.
- Safari 5.1, 8: The print dialog always appears.
- IE 9, 10, 11, Edge 0.11: The print dialog always appears.

Rob--W committed:
Add parseDoc and onParsedDoc to the PDF manager

Move the logic of retrieving a parsed PDF from core/worker.js to
core/pdf_manager.js. This allows other APIs to get a parsed PDF without
being responsible for its creation.

Rob--W committed:
Remove superfluous third parameter from .then

Rob--W committed:
Chromium: Use PDFium for printing if available.

Currently, printing in PDF.js for Chromium is implemented as follows:

1. Render all PDF pages on a <canvas>
2. Send a print request to the browser.

When Chromium receives the print request, it will convert the page to
PDF using Skia, and then open a print preview with PDFium (Chrome's
default PDF reader). This method is suboptimal, since it is essentially
PDF -> HTML and HTML -> PDF.

To enhance the performance, PDF.js will not render the PDF, but send the
PDF to the built-in PDF handler for printing when availability of PDFium
is detected. This brings the following advantages:

- The print job will be started much faster, because PDF.js does not
  have to render the PDF file before triggering the print request.

- Large PDF files can now be printed without out-of-memory errors.

- The generated prints will generally have a better quality (no page
  cut-offs, etc.), because the browser has more control over printer
  settings than PDF.js.

This feature is enabled by default for the GENERIC build, and can
completely be disabled by setting WITH_PRINTING to false. This is also
the first commit that uses editing/pdf_data_writer.js, which was
introduced in an earlier commit.

xlc commented on make.js:
```diff
@@ -128,7 +129,7 @@ target.generic = function() {
   mkdir('-p', GENERIC_DIR + '/web/cmaps');
 
-  var defines = builder.merge(DEFINES, {GENERIC: true});
+  var defines = builder.merge(DEFINES, {GENERIC: true, WITH_EDITING: true});
```
Setting `WITH_EDITING` to true here is too late.
`target.bundle({});` called previously already checking `WITH_EDITING` and see it is false and excluding `editing/` files

Rob--W replied:
Good catch, thanks.

xlc commented:
Doesn't work for password protected file
Here is a one: https://drive.google.com/file/d/0B6YhVF9p-y2AY3hjV2tDcHVWRGs/view?usp=sharing

xlc commented:
On OS X, <kbd>cmd</kbd>+<kbd>P</kbd> doesn't work any more. It prints without preprocess the file. However <kbd>ctrl</kbd>+<kbd>P</kbd> works.

xlc commented on web/viewer.js:
```diff
@@ -1885,6 +1984,23 @@ window.addEventListener('keydown', function keydown(evt) {
+//#endif
+        if (shouldHandlePKey) {
+          PDFViewerApplication.requestPrint();
+          evt.stopImmediatePropagation();
```
need `evt.preventDefault();` to stop triggering bowser's default printing behaviour

Rob--W replied:
I didn't see two print dialogs in Chrome. Nevertheless, I added preventDefault.

xlc commented on web/viewer.js:
```diff
@@ -590,6 +586,109 @@ var PDFViewerApplication = {
+      // Given the lack of alternatives, this is the best that we can get in
+      // terms of feature detection.
+      testEmbed.addEventListener('message', function() {
+        clearTimeout(noReplyAssumeNotAvailable);
```
During my testing, this callback sometimes get called after `noReplyAssumeNotAvailable` triggered especially for large file. In such case means both `yes` and `no` callback are called.
Need to make sure the `yes` callback is not called after timeout.

Rob--W replied:
Looks like a race condition. I've fixed this by adding a boolean guard flag.

xlc commented:
This will generate invalid PDF for printing for some file.
Original file: https://drive.google.com/file/d/0B6YhVF9p-y2AQnJkN0NMdlJyZEk/view?usp=sharing
Generated file: https://drive.google.com/file/d/0B6YhVF9p-y2AWWNwMkF5QzNSeFk/view?usp=sharing

All PDF readers I have (Preview, Adobe Acrobat Reader DC, and even Google Drive preview) says the generated one is invalid file. However PDF.js happily render it and auto-print on load.
iText RUPS say "The document has no page root (meaning: it's an invalid PDF)."

Rob--W commented:
> On OS X, <kbd>cmd</kbd>+<kbd>P</kbd> doesn't work any more. It prints with the browser default behaviour. However <kbd>ctrl</kbd>+<kbd>P</kbd> works.

Weird. I don't thread Cmd different from Ctrl. Is the difference observable even after adding `event.preventDefault()` as you've suggested before?

> Doesn't work for password protected file
> Here is one: https://drive.google.com/file/d/0B6YhVF9p-y2AY3hjV2tDcHVWRGs/view?usp=sharing
I think we should just use the canvas printing method for password protected file.

Sounds like a good plan. I might add support for password-protected PDFs in a later PR, if feasible.

> This will generate invalid PDF for printing for some file.
Original file: https://drive.google.com/file/d/0B6YhVF9p-y2AQnJkN0NMdlJyZEk/view?usp=sharing
Generated file: https://drive.google.com/file/d/0B6YhVF9p-y2AWWNwMkF5QzNSeFk/view?usp=sharing
>
All PDF readers I have (PDFium in Chrome, Preview, Adobe Acrobat Reader DC, and even Google Drive preview) report the generated one is invalid. However PDF.js happily render it and auto-print on load.
iText RUPS says "The document has no page root (meaning: it's an invalid PDF)."

Apparently hybrid xref tables are not supported for files that originally consisted of XRef streams only... I'll submit a fix for that.

xlc commented:
> Weird. I don't thread Cmd different from Ctrl. Is the difference observable even after adding event.preventDefault() as you've suggested before?
Without `preventDefault`, Cmd+P will trigger Chrome's print function. I can see the File menu flash blue indicate Chrome's shortcut is triggered. And obversely its default print function won't work.
With `preventDefault`, Chrome's print function is not triggered, File menu no longer flash blue, and it prints properly.

Rob--W committed:
Disable fast printing if PDF is encrypted

To modify and print an encrypted PDF, we have to decrypt the full PDF
first. This is currently not supported yet.

Rob--W commented:
Fixed all mentioned issues, ready for another round of review:

- Password-protected PDFs are ignored and printed using the current method.
- PDFs that exclusively use XRef streams are also printed correctly.
- Ctrl/Cmd + P calls `event.preventDefault()` when needed.

Future plans:

- Support printing of encrypted PDFs using the new method.
- Add a setting (Chrome extension) or another UI control to allow users to always print using PDF.js (= the current method).

xlc commented:
I found out you can `postMessage` to PDFium to give it password and ask for print

```
diff --git a/web/viewer.js b/web/viewer.js
index 2c10e0a..9c34b03 100644
--- a/web/viewer.js
+++ b/web/viewer.js
@@ -608,10 +608,34 @@ var PDFViewerApplication = {
     this._printAction = printAction;
 
     canUsePDFium(function yes() {
-      PDFViewerApplication.pdfDocument.getDataForPrinting().then(function(res) {
-        printAction.pdfBlobUrl = res.pdfBlobUrl;
+      PDFViewerApplication.pdfDocument.getData().then(function(data) {
+        var blob = PDFJS.createBlob(data, 'application/pdf');
+        var blobUrl = URL.createObjectURL(blob);
+        printAction.pdfBlobUrl = blobUrl;
         printAction.embed = createPDFiumEmbed();
         printAction.embed.src = printAction.pdfBlobUrl;
+        printAction.embed.addEventListener('message', function(evt) {
+          var data = evt.data;
+          console.log(data);
+          switch (data.type) {
+            case 'loadProgress':
+              // TODO: show load progress in a progress bar
+              if (data.progress === 100) { // load successful
+                printAction.embed.postMessage({
+                  type: 'print'
+                });
+              } else if (data.progress === -1) { // load failed
+                onFallback()
+              }
+              break;
+            case 'getPassword':
+              printAction.embed.postMessage({
+                type: 'getPasswordComplete',
+                password: 'password' // TODO: ask user for password or get it from worker
+              });
+              break;
+          }
+        });
         document.body.appendChild(printAction.embed);
         // Clean up when the print dialog is closed.
         window.addEventListener('focus', printAction);
```

xlc commented on web/viewer.js:
```diff
@@ -590,6 +586,112 @@ var PDFViewerApplication = {
+    // and usable without click-to-play. The result is not cached because the
+    // user can change the value via the browser settings.
+    // This feature detection never has false positives, only false negatives.
+    function canUsePDFium(yes, no) {
```
`canUsePDFium` should return a promise and resolve/reject it instead of yes/no callbacks.

xlc commented:
embed tag with type of `application/x-google-chrome-pdf` no longer works in Chrome v46.

Rob--W commented:
@xlc And it does not work in Chrome 45 either. We have to use the default MIME-type (`application/pdf`) instead, and include proper guards to ensure that we don't accidentally let PDF.js render it again (which is why I was using this MIME-type in the first place).

timvandermeij commented:
Closing since this is not in a mergeable state and we should strive to find a solution that works for all browsers (refer to the mentioned issues in https://github.com/mozilla/pdf.js/issues/8007#issuecomment-277714270). We'll keep on tracking this issue in e.g., #8007 and others.
